component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/WSO2TokenCredential.java

```diff
@@ -50,11 +51,28 @@ public WSO2TokenCredential(String authEndpoint, String authToken) {
             String sasToken = AuthClient.getSASToken(this.authEndpoint, this.authToken);
             log.debug("New SAS token retrieved.");
             // Using lower duration than actual.
-            OffsetDateTime time = OffsetDateTime.now(ZoneOffset.UTC).plus(Duration.ofHours(20));
+            OffsetDateTime time = getExpirationTime(sasToken);
             return Mono.fromCallable(() -> new AccessToken(sasToken, time));
         } catch (AuthenticationException e) {
             log.error("Error occurred when retrieving SAS token.", e);
             throw new RuntimeException("Error occurred when retrieving SAS token.", e);
         }
     }
+
+    private OffsetDateTime getExpirationTime(String sharedAccessSignature) {
+        String[] parts = sharedAccessSignature.split("&");
+        return Arrays.stream(parts).map(part -> part.split("="))
+                .filter(pair -> pair.length == 2 && pair[0].equalsIgnoreCase("se"))
+                .findFirst().map(pair -> pair[1])
+                .map((expirationTimeStr) -> {
+                    try {
+                        long epochSeconds = Long.parseLong(expirationTimeStr);
+                        return Instant.ofEpochSecond(epochSeconds).atOffset(ZoneOffset.UTC);
+                    } catch (NumberFormatException e) {
+                        log.error("Invalid expiration time format in the SAS token.", e);
+                        return OffsetDateTime.MAX;
+                    }
+                })
+                .orElse(OffsetDateTime.MAX);
```

Comment:
shouldn't we set a default instead of providing an unexpired token?

Reply:
actually, this is the logic defined in the azure default token credentials. I think even if we set it to default or max there will be an authentication failure and then again it will go in retry flow.

Reply:
oh ok.